### PR TITLE
Stream parsed content to disk

### LINKED ISSUE #11: Switch to streaming files to disk
1. Change `uids_to_docs` to be a generator. This means removing the `return` and sticking a `yield`  inside the for loop (see [here](https://realpython.com/introduction-to-python-generators/)).
2. Move `uids_to_docs` into the for loop in `main.py`. This is what will cause iteration.
3. Test.

### PR DESCRIPTION
**Pull request by Anwesh1**
# Overview

Right now, I'm changing the last value of the list returned from ```get_list_pmid```  to "bruh" and in the ```uids_to_docs``` I'm checking if "bruh" is present, and if yes to raise an exception. 
The exception is being raised but the way we were hoping for the ```yield``` to work it isn't doing that. Probably need to change the way the helper functions are using the list of PMID's

## Closes

Closes #11.

### CODE REVIEW

**Comment by JohnGiorgi** [comment]
Can you show an example of you calling the program, the expected output, and the actual output? (as concisely as possible please!)

File: pubmed_dl/ncbi.py [hunk 1]
```diff
@@ -112,28 +113,31 @@ def uids_to_docs(uids: List[str]) -> List[Dict[str, Collection[Any]]]:
...
+                raise e
             eutil_response = _get_eutil_records("efetch", id, rettype="medline", retmode="text")
         except Exception as e:
             print(f"Error encountered in uids_to_docs {e}")
             raise e
         else:
             output = _medline_to_docs(eutil_response)
             docs = docs + output
-    return docs
+    yield docs
```
Comment (jvwong) [inline pubmed_dl/ncbi.py]: I think you want something more like:

```
for i in range(num_queries):
        lower = i * MAX_EFETCH_RETMAX
        ...
        try:
            ...
        else:
            yield output
```

Comment (JohnGiorgi) [reply pubmed_dl/ncbi.py]: Yup the yield should be inside the for loop

**Comment by Anwesh1** [comment]
@JohnGiorgi 
So you can run the ```main.py``` on the  ```PMID-retrieval-updated``` branch. The expected output is in the format of:
```json
{"uid": "PMID here", "text": "Title and abstract here}
```
This should output 12202 records (because the 12203 is the "faulty" record). However, nothing is being written to the file.

**Comment by JohnGiorgi** [comment]
> @JohnGiorgi
> So you can run the `main.py` on the `PMID-retrieval-updated` branch. The expected output is in the format of:
> 
> ```json
> {"uid": "PMID here", "text": "Title and abstract here}
> ```
> 
> This should output 12202 records (because the 12203 is the "faulty" record). However, nothing is being written to the file.

See @jvwong 's fix above. Let me know if the error persists after that fix. PS it may be useful to [brush up on generators briefly](https://realpython.com/introduction-to-python-generators/#example-1-reading-large-files).

File: pubmed_dl/ncbi.py (hunk 1 above)
Comment (Anwesh1) [reply pubmed_dl/ncbi.py]: Perfect!! This works.

**Comment by Anwesh1** [comment]
So @jvwong fix works perfectly now. It's like we expected, it is done in batches, so in this case, the first 10,000 were good, however, the next batch didn't output (which is fine, worst case we lose 1 batch).
Right now all the data is printing on 1 line, do you want that format or would you like every record on a separate line?

**Comment by JohnGiorgi** [comment]
Sorry are you saying we don't output the last batch or just if something goes wrong with the pipeline we at worst loose the last batch?

All records should be on their own line. Use f-string formatting and the "\n" character to achieve this.

File: pubmed_dl/ncbi.py [hunk 2]
```diff
@@ -112,28 +113,31 @@ def uids_to_docs(uids: List[str]) -> List[Dict[str, Collection[Any]]]:
         upper = min([lower + MAX_EFETCH_RETMAX, num_uids])
         id = uids[lower:upper]
         try:
+            if "bruh" in id:
+                raise e
             eutil_response = _get_eutil_records("efetch", id, rettype="medline", retmode="text")
```
Comment (jvwong) [inline pubmed_dl/ncbi.py:115]: `_get_eutil_records` is the function that will tend to fail, if anything. Right now it's passing along the error but we could change it to continue on e.g. literally log error and `continue`

**Comment by Anwesh1** [comment]
Yeah what @jvwong said essentially. So let's say there's 25,000 records and the error is come across on the 23,000th record then we're still going to have the data of the first 20,000 just the 20,000 to 30,000 batch is lost. Which is totally fine.

**Comment by jvwong** [comment]
What if the problem is record 1 or 10...

**Comment by Anwesh1** [comment]
@jvwong then we have a problem essentially. The whole point of this is so that we don't run it for 5 hours then come across a record thats's problematic then we lose all the 5 hours worth. However, now we will have the data of the 5 hours and then have a general idea of where there was a problem because that batch and every batch after it won't be outputted.

**Comment by jvwong** [comment]
> @jvwong then we have a problem essentially. The whole point of this is so that we don't run it for 5 hours then come across a record thats's problematic then we lose all the 5 hours worth. However, now we will have the data of the 5 hours and then have a general idea of where there was a problem because that batch and every batch after it won't be outputted.

I'm claiming that given this error (record 1 or 10), the code in this PR will result in 0 records being saved. Is that true?

**Comment by Anwesh1** [comment]
yes, if the error is in the 1st 10,000 records we will have nothing saved.

**Comment by jvwong** [comment]
Are you able to update it so it can recover from any error and complete the 'job'

**Comment by Anwesh1** [comment]
I think if we were to sort of change the code to basically look over a set of 10,000 if there is an error in them. Like what you said earlier, let's say an exception is thrown then we forget about that batch and ```continue``` onto the next batch. That should be possible in my understanding.

**Comment by jvwong** [comment]
Sounds good. Try it out then push the code.

**Comment by Anwesh1** [comment]
So I can just add a ```continue``` here (because this essentially captures all the errors)
https://github.com/PathwayCommons/pubmed-dl/blob/1e40bab1d1d9ecac46c0afbeb27a2b8cfe6afce1/pubmed_dl/ncbi.py#L119-L121
If we have a ```continue``` there, we can't have a ```raise``` though. It will just print that an error occurred but an exception won't be raised. I think that is fine, what do you guys think?

**Comment by JohnGiorgi** [comment]
> So I can just add a `continue` here (because this essentially captures all the errors)
> 
> https://github.com/PathwayCommons/pubmed-dl/blob/1e40bab1d1d9ecac46c0afbeb27a2b8cfe6afce1/pubmed_dl/ncbi.py#L119-L121
> 
> If we have a `continue` there, we can't have a `raise` though. It will just print that an error occurred but an exception won't be raised. I think that is fine, what do you guys think?

Why don't we leave this PR at just modifying the code to write to the file in batches, and open another one for dealing with errors? I feel that these PRs are generally too big and are slowing us down. Open an issue to track our fix for handling exceptions

**Comment by Anwesh1** [comment]
Okay, so the batches are done, I'm just trying to figure out where to put the ```\n``` character so everything prints on a newline.

File: pubmed_dl/main.py [hunk 3]
```diff
@@ -1,18 +1,23 @@
...
+    end = "2020/12/07"
     val = get_list_pmid(start,end)
+    val[-1] = "bruh"
     here = uids_to_docs(val)
-    for item in here:
-        ab.write(json.dumps(item))
-        ab.write("\n")
+    for item in uids_to_docs(val):
+        ab.write(json.dumps(item) + '\n')
+
```
Comment (JohnGiorgi) [inline pubmed_dl/main.py]:
```python
for batch in uids_to_docs(val):
    for doc in batch:
        ab.write(f"{json.dumps(doc)}\n")
```

Should fix your issue of writing an entire batch to one line. Test it though

File: pubmed_dl/ncbi.py [hunk 4]
```diff
@@ -100,7 +101,7 @@ def _medline_to_docs(records: List[Dict[str, str]]) -> List[Dict[str, Collection
         docs.append({"uid": pmid, "text": text})
     return docs
 
-
+#change to generator
```
Comment (JohnGiorgi) [inline pubmed_dl/ncbi.py]: Drop

File: pubmed_dl/ncbi.py [hunk 5]
```diff
@@ -112,28 +113,31 @@ def uids_to_docs(uids: List[str]) -> List[Dict[str, Collection[Any]]]:
         upper = min([lower + MAX_EFETCH_RETMAX, num_uids])
         id = uids[lower:upper]
         try:
+            if "bruh" in id:
+                raise e
```
Comment (JohnGiorgi) [inline pubmed_dl/ncbi.py]: Drop before committing

**Review by JohnGiorgi — CHANGES_REQUESTED** [review]
Before you comit, will you

1. Run `pip install -e ".[dev]"`
2. Run `black .` inside `pubmed-dl`
3. Run `flake8 .` inside `pubmed-dl`. Fix any reported errors

File: pubmed_dl/ncbi.py (hunk 1 above)
Comment (JohnGiorgi) [reply pubmed_dl/ncbi.py]: @Anwesh1 Make sure to actually commit the change

File: pubmed_dl/main.py (hunk 3 above)
Comment (Anwesh1) [reply pubmed_dl/main.py]: This works thanks

File: pubmed_dl/main.py [hunk 6]
```diff
@@ -1,18 +1,23 @@
...
 def main():
     ab = open("test.json", 'w')
-    start = input("Enter start date (YYYY/MM/DD): ")
-    end = input("Enter end date (YYYY/MM/DD): ")
+    #start = input("Enter start date (YYYY/MM/DD): ")
+    #end = input("Enter end date (YYYY/MM/DD): ")
+    start = "2020/12/04"
+    end = "2020/12/07"
     val = get_list_pmid(start,end)
+    val[-1] = "bruh"
```
Comment (JohnGiorgi) [inline pubmed_dl/main.py]: I would think almost all of this should be reverted before committing, looks like it's related to your testing. In the near-future we should move testing _outside_ the actual code.

**Comment by Anwesh1** [comment]
I haven't really worked on this using the command line. so am i supposed to do the steps on the ```README.md```? Cloning then installing?

**Comment by JohnGiorgi** [comment]
> I haven't really worked on this using the command line. so am i supposed to do the steps on the `README.md`? Cloning then installing?

I am not sure what you mean. You are already working on a local clone of this repository. You should commit each change locally then push it here (you have already been doing this)

**Comment by Anwesh1** [comment]
I'm trying to run the ```pip install -e ".[dev]"``` command but my command line is erroring out for some reason. So I re-cloned the repo and tried running the ```pip install --editable``` and that is erroring out too. It might just be windows bugging

**Comment by JohnGiorgi** [comment]
> I'm trying to run the `pip install -e ".[dev]"` command but my command line is erroring out for some reason. So I re-cloned the repo and tried running the `pip install -editable` and that is erroring out too. It might just be windows bugging

Can you paste the output? Try to always do this, otherwise, I will be of almost no help.

`pip install -e` is wrong though. You need a file path (e.g. `pip install -e .`). But we specifically need the `[dev]` to install `black` and `flake8`. For now, you can just `pip install black` and `pip install flake8` to keep this moving

**Comment by Anwesh1** [comment]
This is the error:
image
is my command wrong syntactically?

**Comment by JohnGiorgi** [comment]
> This is the error:
> image
> is my command wrong syntactically?

Uh no syntax looks fine. Is this a bash shell? Are you able to pip install other packages? If so, I would just `pip install black flake8` for now so we can move on

**Comment by Anwesh1** [comment]
I just did ```pip install black flake8``` they seemed to download fine.
This is the Command line for windows.

**Comment by JohnGiorgi** [comment]
> i just did `pip install black flake8` they seemed to download fine.

Great. I only see one commit here. Can you commit all the changes for the comments you have resolved and the remaining comments, then perform those three steps with `black` and `flake8`. Once all the tests pass let's merge.

**Comment by jvwong** [comment]
> Before you comit, will you
> 
>     1. Run `pip install -e ".[dev]"`
> 
>     2. Run `black .` inside `pubmed-dl`
> 
>     3. Run `flake8 .` inside `pubmed-dl`. Fix any reported errors

This works for me on Mac/python 3.8

**Comment by Anwesh1** [comment]
My PC's internet connection just crashed. The internet is working for every device except for my PC (the DNS server isn't responding apparently) so I'm trying to fix that. I apologize for any delays. Sent this from my phone.

**Comment by Anwesh1** [comment]
Hey @JohnGiorgi @jvwong I'm sorry, my computer literally forgot how to connect to the internet. I had to reset it so I had to re-download Python and all required packages and stuff. So I finally started it again and I tried running the files after our earlier updates to write in batches. I'm coming across this error:
image
I'm committing the updated files, could you please run them and check if they work? It maybe because I might be missing something after I reset the computer. Thanks again.

**Comment by JohnGiorgi** [comment]
You committed your changes to `main`, not `PMID-retrieval-updated`, I can fix it though and run the formatted and linter.

**Comment by Anwesh1** [comment]
@JohnGiorgi Thanks for that, I reverted it back to the original main. Thanks for merging to the ```PMID-retrieval-updated``` branch.
Does it give you an error when you run?

**Comment by JohnGiorgi** [comment]
> @JohnGiorgi Thanks for that, I reverted it back to the original main. Thanks for merging to the `PMID-retrieval-updated` branch.
> Does it give you an error when you run?

Yes, here is my I/O. Looks like an ill-formed URL/request but not sure.

```bash
Enter start date (YYYY/MM/DD): 2021/01/01
Enter end date (YYYY/MM/DD): 2021/30/01
Total records: 192731
URL 1: https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?&retmax=100000&query_key=1&db=pubmed&rettype=uilist&retmode=text&WebEnv=MCID_601c8527a188e068371a84ae&retstart=0
URL 2: https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?&retmax=100000&query_key=1&db=pubmed&rettype=uilist&retmode=text&WebEnv=MCID_601c8527a188e068371a84ae&retstart=100000
HTTP error 400 Client Error: Bad Request for url: https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi; status code: 400
Error encountered in uids_to_docs 400 Client Error: Bad Request for url: https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi
Traceback (most recent call last):
  File "/Users/johngiorgi/miniconda3/envs/pubmed-dl/bin/pubmed-dl", line 33, in <module>
    sys.exit(load_entry_point('pubmed-dl', 'console_scripts', 'pubmed-dl')())
  File "/Users/johngiorgi/Documents/dev/pubmed-dl/pubmed_dl/main.py", line 11, in main
    for batch in uids_to_docs(val):
  File "/Users/johngiorgi/Documents/dev/pubmed-dl/pubmed_dl/ncbi.py", line 119, in uids_to_docs
    raise e
  File "/Users/johngiorgi/Documents/dev/pubmed-dl/pubmed_dl/ncbi.py", line 116, in uids_to_docs
    eutil_response = _get_eutil_records("efetch", id, rettype="medline", retmode="text")
  File "/Users/johngiorgi/Documents/dev/pubmed-dl/pubmed_dl/ncbi.py", line 87, in _get_eutil_records
    eutilResponse = _safe_request(url, "POST", files=eutils_params)
  File "/Users/johngiorgi/Documents/dev/pubmed-dl/pubmed_dl/ncbi.py", line 48, in _safe_request
    r.raise_for_status()
  File "/Users/johngiorgi/miniconda3/envs/pubmed-dl/lib/python3.8/site-packages/requests/models.py", line 943, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi
```

**Comment by Anwesh1** [comment]
Oh I see, it's weird that the exact same piece of code was working perfectly. Let me try to fix it.

**Comment by Anwesh1** [comment]
@jvwong  The bad request for URL is caused apparently because the mandatory parameter db is omitted. But doesn't the ```files=eutils_params``` get the ```db``` in these lines?
https://github.com/PathwayCommons/pubmed-dl/blob/99ed379e612464a48cbf7976f2d337db040b026a/pubmed_dl/ncbi.py#L72-L86
I'd appreciate your input if possible

**Comment by jvwong** [comment]
> @jvwong The bad request for URL is caused apparently because the mandatory parameter db is omitted. But doesn't the `files=eutils_params` get the `db` in these lines?
> 
> https://github.com/PathwayCommons/pubmed-dl/blob/99ed379e612464a48cbf7976f2d337db040b026a/pubmed_dl/ncbi.py#L72-L86
> 
> I'd appreciate your input if possible

Two things:

1. You need to initialize the value for `NCBI_EUTILS_API_KEY`. Also this is commented out in the `.env`.
2. Weird error: Apparently, one of the PMIDs provided to the code doesn't exist so throws a key error in  `_medline_to_docs`
`id: 33528541 Error occurred: The following PMID is not available: 33528541`

**Comment by Anwesh1** [comment]
Hey @jvwong thanks for that, can't believe I didn't see that. Right now the code works for most of the dates. However, there is an error in the "PMID" keyError:
```bash
Traceback (most recent call last):
  File "c:\Users\anwes\Desktop\pupul\BaderLab\Pubmed\semantic_git\pubmed-dl\pubmed_dl\main.py", line 17, in <module>
    main()
  File "c:\Users\anwes\Desktop\pupul\BaderLab\Pubmed\semantic_git\pubmed-dl\pubmed_dl\main.py", line 10, in main
    for batch in uids_to_docs(val):
  File "c:\Users\anwes\Desktop\pupul\BaderLab\Pubmed\semantic_git\pubmed-dl\pubmed_dl\ncbi.py", line 120, in uids_to_docs
    output = _medline_to_docs(eutil_response)
  File "c:\Users\anwes\Desktop\pupul\BaderLab\Pubmed\semantic_git\pubmed-dl\pubmed_dl\ncbi.py", line 97, in _medline_to_docs
    pmid = record["PMID"]
KeyError: 'PMID'
```
I get that when I run it for the PMID's on the 2nd of this month.

**Comment by Anwesh1** [comment]
There is also something that we're doing wrong with the printing aspect, this one:
```python
with open("test.json", "w") as f:
    for batch in uids_to_docs(val):
         for doc in batch:
              f.write(f"{json.dumps(doc)}\n")
```
I even tried going back to how we were doing it originally:
```python
for item in uids_to_docs(val):
        ab.write(f"{json.dumps(item)}\n")
```
There are multiple entries for the same record for some reason. I think the iteration is piling on to data from previous iterations. I've spent a few hours on this but can't seem to figure out why that could be happening. I would appreciate if someone could take a look when free. Thanks.

**Comment by JohnGiorgi** [comment]
> There is also something that we're doing wrong with the printing aspect, this one:
> 
> ```python
> with open("test.json", "w") as f:
>     for batch in uids_to_docs(val):
>          for doc in batch:
>               f.write(f"{json.dumps(doc)}\n")
> ```
> 
> I even tried going back to how we were doing it originally:
> 
> ```python
> for item in uids_to_docs(val):
>         ab.write(f"{json.dumps(item)}\n")
> ```
> 
> There are multiple entries for the same record for some reason. I think the iteration is piling on to data from previous iterations. I've spent a few hours on this but can't seem to figure out why that could be happening. I would appreciate if someone could take a look when free. Thanks.

What’s the issue with the first solution? I can test later but I would encourage you to work through it.

**Comment by Anwesh1** [comment]
Hey @JohnGiorgi I feel bad coming to you guys repeatedly, so sorry about that.
 I'm trying to fix it but essentially, with the first solution let's say there are 50,000 records. It is printing the 1st 10,000 then it is printing the 1st 20,000 (inclusive of the 1st 10,000 it already printed) then it is printing the first 30,000 and it continues like that. So 50,000 entries are actually printing out about 150,000 records where the first 10,000 are repeated 5 times and the next batch is repeated 4 times and so on. That's basically the problem.

**Comment by Anwesh1** [comment]
Hey, @JohnGiorgi got it working now (finally :) ), you can test the committed files and let me know so I can merge. Thanks for the help.

**Review by JohnGiorgi — APPROVED** [review]

**Pull request merged**